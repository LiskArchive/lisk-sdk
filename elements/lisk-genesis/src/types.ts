/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { AccountSchema, Account, GenesisBlockHeader } from '@liskhq/lisk-chain';

export type PartialReq<T, Keys extends keyof T = keyof T> = Pick<
	Partial<T>,
	Exclude<keyof T, Keys>
> &
	{
		[K in Keys]: T[K];
	};

export type GenesisBlockHeaderWithoutId = Omit<GenesisBlockHeader, 'id'>;

export interface GenesisBlockParams {
	// List of accounts in the genesis
	readonly accounts: ReadonlyArray<Account>;
	// List fo initial delegate addresses used during the bootstrap period to forge blocks
	readonly initDelegates: ReadonlyArray<Buffer>;
	// Number of blocks per round
	readonly roundLength: number;
	// Account Schema for the genesis block
	readonly accountAssetSchemas: { [moduleName: string]: AccountSchema };
	// Number of rounds for bootstrap period, default is 3
	readonly initRounds?: number;
	readonly height?: number;
	readonly timestamp?: number;
	readonly previousBlockID?: Buffer;
}
