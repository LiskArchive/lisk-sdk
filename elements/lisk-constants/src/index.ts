/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
// tslint:disable-next-line:no-magic-numbers
export const EPOCH_TIME = new Date(Date.UTC(2016, 4, 24, 17, 0, 0, 0));
export const EPOCH_TIME_MILLISECONDS = EPOCH_TIME.getTime();
const MS_FACTOR = 1000;
export const EPOCH_TIME_SECONDS = Math.floor(EPOCH_TIME.getTime() / MS_FACTOR);

// Largest possible number which can be stored in eight bytes.
// Derived from bignum.fromBuffer(Buffer.from(new Array(8).fill(255))).
const MAX_EIGHT_BYTE_NUMBER = '18446744073709551615';

export const MAX_ADDRESS_NUMBER = MAX_EIGHT_BYTE_NUMBER;
export const MAX_TRANSACTION_ID = MAX_EIGHT_BYTE_NUMBER;
// Largest possible amount. Maximum value for PostgreSQL bigint.
export const MAX_TRANSACTION_AMOUNT = '9223372036854775807';

export const SIGNED_MESSAGE_PREFIX = 'Lisk Signed Message:\n';

export const TESTNET_NETHASH =
	'da3ed6a45429278bac2666961289ca17ad86595d33b31037615d4b8e8f158bba';
export const MAINNET_NETHASH =
	'ed14889723f24ecc54871d058d98ce91ff2f973192075c0155ba2b7b70ad2511';
