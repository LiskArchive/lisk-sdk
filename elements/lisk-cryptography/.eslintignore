dist-node
jest.config.js
types
benchmark
.eslintrc.js
test/_setup.js
