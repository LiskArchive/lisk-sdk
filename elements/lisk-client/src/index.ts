/*
 * Copyright Â© 2018 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { APIClient as APIClientModule } from '@liskhq/lisk-api-client';
import * as constantsModule from '@liskhq/lisk-constants';
import * as cryptographyModule from '@liskhq/lisk-cryptography';
import * as passphraseModule from '@liskhq/lisk-passphrase';
import * as transactionModule from '@liskhq/lisk-transactions';

// tslint:disable-next-line variable-name
export const APIClient = APIClientModule;
export const constants = constantsModule;
export const cryptography = cryptographyModule;
export const passphrase = passphraseModule;
export const transaction = transactionModule;

// tslint:disable-next-line no-default-export
export default {
	APIClient,
	constants,
	cryptography,
	passphrase,
	transaction,
};
