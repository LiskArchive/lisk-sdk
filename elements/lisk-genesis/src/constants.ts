/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { hash } from '@liskhq/lisk-cryptography';

export const EMPTY_BUFFER = Buffer.alloc(0);
export const EMPTY_HASH = hash(EMPTY_BUFFER);

export const GENESIS_BLOCK_VERSION = 0;
export const GENESIS_BLOCK_GENERATOR_PUBLIC_KEY = EMPTY_BUFFER;
export const GENESIS_BLOCK_REWARD = BigInt(0);
export const GENESIS_BLOCK_SIGNATURE = EMPTY_BUFFER;
export const GENESIS_BLOCK_TRANSACTION_ROOT = EMPTY_HASH;
export const GENESIS_BLOCK_MAX_BALANCE = BigInt(2) ** BigInt(63) - BigInt(1);
