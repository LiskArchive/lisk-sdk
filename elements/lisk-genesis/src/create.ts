/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec, Schema } from '@liskhq/lisk-codec';
import { hash } from '@liskhq/lisk-cryptography';
import {
	getAccountSchemaAndDefault,
	getGenesisBlockHeaderAssetSchema,
	Account,
	GenesisBlock,
	blockHeaderSchema,
} from '@liskhq/lisk-chain';
import { objects } from '@liskhq/lisk-utils';
import {
	EMPTY_BUFFER,
	GENESIS_BLOCK_GENERATOR_PUBLIC_KEY,
	GENESIS_BLOCK_REWARD,
	GENESIS_BLOCK_SIGNATURE,
	GENESIS_BLOCK_TRANSACTION_ROOT,
	GENESIS_BLOCK_VERSION,
} from './constants';
import { GenesisBlockHeaderWithoutId, GenesisBlockParams } from './types';

const getBlockId = (header: GenesisBlockHeaderWithoutId, accountSchema: Schema): Buffer => {
	// eslint-disable-next-line
	const genesisBlockAssetBuffer = codec.encode(
		getGenesisBlockHeaderAssetSchema(accountSchema),
		header.asset,
	);

	const genesisBlockHeaderBuffer = codec.encode(blockHeaderSchema, {
		...header,
		asset: genesisBlockAssetBuffer,
	});

	return hash(genesisBlockHeaderBuffer);
};

const createAccount = <T>(account: Account, defaultAccount: Record<string, unknown>): Account =>
	objects.mergeDeep({}, objects.cloneDeep(defaultAccount), account) as Account<T>;

export const createGenesisBlock = (params: GenesisBlockParams): GenesisBlock => {
	// Default values
	const initRounds = params.initRounds ?? 3;
	const height = params.height ?? 0;
	const timestamp = params.timestamp ?? Math.floor(Date.now() / 1000);
	const previousBlockID = params.previousBlockID ?? Buffer.from(EMPTY_BUFFER);
	const { schema: accountSchema, defaultAccount } = getAccountSchemaAndDefault(
		params.accountAssetSchemas,
	);

	// Constant values
	const version = GENESIS_BLOCK_VERSION;
	const generatorPublicKey = GENESIS_BLOCK_GENERATOR_PUBLIC_KEY;
	const reward = GENESIS_BLOCK_REWARD;
	const signature = GENESIS_BLOCK_SIGNATURE;
	const transactionRoot = GENESIS_BLOCK_TRANSACTION_ROOT;

	const accounts: ReadonlyArray<Account> = params.accounts
		.map(account => createAccount(account, defaultAccount))
		.sort((a, b): number => a.address.compare(b.address));

	const initDelegates: ReadonlyArray<Buffer> = [...params.initDelegates].sort((a, b): number =>
		a.compare(b),
	);

	const header: GenesisBlockHeaderWithoutId = {
		generatorPublicKey,
		height,
		previousBlockID,
		reward,
		signature,
		timestamp,
		transactionRoot,
		version,
		asset: {
			initRounds,
			initDelegates,
			accounts,
		},
	};

	const genesisBlock: GenesisBlock = {
		header: {
			...header,
			id: getBlockId(header, accountSchema),
		},
		payload: [],
	};

	return genesisBlock;
};
