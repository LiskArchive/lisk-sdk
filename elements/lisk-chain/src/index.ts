/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK } from './constants';

const events = { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK };

export { events };
export { Chain } from './chain';
export { Transaction, transactionSchema } from './transaction';
export {
	blockHeaderAssetSchema,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	validatorsSchema,
} from './schema';
export type {
	Account,
	AccountDefaultProps,
	RawBlock,
	RawBlockHeader,
	GenesisBlock,
	GenesisBlockHeader,
	Block,
	BlockHeader,
	Validator,
	AccountSchema,
} from './types';
export { readGenesisBlockJSON } from './utils';

export type { StateStore } from './state_store';
export type { Slots } from './slots';
