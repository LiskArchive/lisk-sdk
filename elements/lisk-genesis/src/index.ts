/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { createGenesisBlock } from './create';

export * from './types';
export { createGenesisBlock };
