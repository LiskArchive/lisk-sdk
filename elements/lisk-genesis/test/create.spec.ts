/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { hash, getRandomBytes } from '@liskhq/lisk-cryptography';
import { objects } from '@liskhq/lisk-utils';
import { Account } from '@liskhq/lisk-chain';
import { createGenesisBlock } from '../src';
import { validGenesisBlockParams } from './fixtures';

describe('create', () => {
	it('should create genesis block', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock).toMatchSnapshot();
	});

	it('should set "version" to zero', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.version).toEqual(0);
	});

	it('should set "reward" to zero', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.reward).toEqual(BigInt(0));
	});

	it('should set "transactionRoot" to empty hash', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.transactionRoot).toEqual(hash(Buffer.alloc(0)));
	});

	it('should set "generatorPublicKey" to empty buffer', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.generatorPublicKey).toEqual(Buffer.alloc(0));
	});

	it('should set "signature" to empty buffer', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.header.signature).toEqual(Buffer.alloc(0));
	});

	it('should set "payload" to empty array', () => {
		// Arrange & Act
		const genesisBlock = createGenesisBlock(validGenesisBlockParams);

		// Assert
		expect(genesisBlock.payload).toEqual([]);
	});

	it('should set "height" to provided height', () => {
		// Arrange
		const height = 10;

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			height,
		});

		// Assert
		expect(genesisBlock.header.height).toEqual(height);
	});

	it('should set "timestamp" to provided timestamp', () => {
		// Arrange
		const timestamp = 1592227157;

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			timestamp,
		});

		// Assert
		expect(genesisBlock.header.timestamp).toEqual(timestamp);
	});

	it('should set "previousBlockID" to provided previousBlockID', () => {
		// Arrange
		const previousBlockID = getRandomBytes(20);

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			previousBlockID,
		});

		// Assert
		expect(genesisBlock.header.previousBlockID).toEqual(previousBlockID);
	});

	it('should set "initRounds" ordering lexicographically', () => {
		// Arrange
		const initDelegates = objects.cloneDeep(validGenesisBlockParams.initDelegates);
		const initDelegatesSorted = objects.cloneDeep(initDelegates) as Buffer[];
		const initDelegatesUnSorted = objects.cloneDeep(initDelegates) as Buffer[];
		initDelegatesSorted.sort((a, b) => a.compare(b));
		initDelegatesUnSorted.sort((a, b) => b.compare(a));

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			initDelegates: initDelegatesUnSorted,
		});

		// Assert
		expect(genesisBlock.header.asset.initDelegates).toEqual(initDelegatesSorted);
	});

	it('should set "accounts" ordering lexicographically by "address"', () => {
		// Arrange
		const accounts = objects.cloneDeep(validGenesisBlockParams.accounts) as Account[];
		accounts.sort((a, b) => b.address.compare(a.address));

		const accountsSortedAddresses = accounts.map(a => a.address).sort((a, b) => a.compare(b));

		// Act
		const genesisBlock = createGenesisBlock({
			...validGenesisBlockParams,
			accounts,
		});

		// Assert
		expect(genesisBlock.header.asset.accounts.map(a => a.address)).toEqual(accountsSortedAddresses);
	});
});
