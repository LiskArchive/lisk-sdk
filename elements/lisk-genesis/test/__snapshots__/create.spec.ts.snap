// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create should create genesis block 1`] = `
Object {
  "header": Object {
    "asset": Object {
      "accounts": Array [
        Object {
          "address": Object {
            "data": Array [
              35,
              112,
              186,
              152,
              69,
              30,
              31,
              240,
              7,
              2,
              159,
              30,
              96,
              46,
              20,
              187,
              9,
              123,
              81,
              111,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 2620126571n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              37,
              212,
              193,
              237,
              45,
              162,
              75,
              168,
              86,
              184,
              148,
              194,
              9,
              24,
              143,
              112,
              224,
              202,
              171,
              110,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 2874239947n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              52,
              179,
              218,
              222,
              213,
              80,
              245,
              176,
              197,
              9,
              71,
              216,
              225,
              119,
              98,
              185,
              115,
              12,
              241,
              249,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 2384412768n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              55,
              43,
              157,
              7,
              109,
              211,
              126,
              124,
              24,
              60,
              246,
              237,
              3,
              193,
              125,
              216,
              205,
              192,
              211,
              18,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 28138131n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              76,
              149,
              19,
              239,
              123,
              11,
              204,
              238,
              2,
              69,
              89,
              71,
              71,
              0,
              170,
              215,
              3,
              119,
              4,
              59,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 476696623n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              82,
              201,
              93,
              36,
              167,
              234,
              67,
              210,
              2,
              218,
              138,
              13,
              167,
              240,
              140,
              51,
              164,
              235,
              109,
              92,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 4218444994n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              98,
              21,
              12,
              207,
              145,
              204,
              22,
              52,
              249,
              48,
              238,
              208,
              183,
              118,
              15,
              66,
              23,
              82,
              166,
              152,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 1340967959n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              110,
              53,
              133,
              177,
              23,
              81,
              66,
              58,
              114,
              235,
              91,
              73,
              245,
              121,
              64,
              190,
              80,
              144,
              71,
              87,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 2165380961n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              128,
              224,
              141,
              50,
              44,
              36,
              181,
              50,
              154,
              37,
              124,
              91,
              27,
              215,
              87,
              128,
              50,
              211,
              185,
              33,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 3116632800n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              157,
              52,
              166,
              149,
              81,
              216,
              184,
              94,
              192,
              245,
              82,
              214,
              94,
              168,
              221,
              53,
              114,
              77,
              238,
              144,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 4056778033n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              192,
              24,
              181,
              221,
              125,
              164,
              182,
              183,
              191,
              78,
              108,
              104,
              142,
              141,
              126,
              152,
              65,
              212,
              170,
              143,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 2910960211n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              195,
              195,
              7,
              23,
              78,
              90,
              227,
              1,
              196,
              148,
              57,
              108,
              15,
              219,
              99,
              7,
              236,
              78,
              198,
              12,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 1966001160n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              195,
              198,
              54,
              21,
              36,
              4,
              51,
              37,
              232,
              171,
              206,
              140,
              193,
              248,
              153,
              226,
              194,
              255,
              217,
              176,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 3756324977n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              217,
              251,
              219,
              79,
              34,
              250,
              166,
              9,
              54,
              160,
              17,
              52,
              87,
              218,
              1,
              158,
              182,
              28,
              61,
              119,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 489340925n,
          },
        },
        Object {
          "address": Object {
            "data": Array [
              222,
              223,
              189,
              76,
              106,
              26,
              209,
              239,
              13,
              122,
              149,
              205,
              105,
              7,
              17,
              69,
              72,
              216,
              198,
              216,
            ],
            "type": "Buffer",
          },
          "dpos": Object {
            "delegate": Object {
              "consecutiveMissedBlocks": 0,
              "isBanned": false,
              "lastForgedHeight": 0,
              "pomHeights": Array [],
              "totalVotesReceived": 0n,
              "username": "",
            },
            "sentVotes": Array [],
            "unlocking": Array [],
          },
          "token": Object {
            "balance": 653021139n,
          },
        },
      ],
      "initDelegates": Array [
        Object {
          "data": Array [
            35,
            112,
            186,
            152,
            69,
            30,
            31,
            240,
            7,
            2,
            159,
            30,
            96,
            46,
            20,
            187,
            9,
            123,
            81,
            111,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            37,
            212,
            193,
            237,
            45,
            162,
            75,
            168,
            86,
            184,
            148,
            194,
            9,
            24,
            143,
            112,
            224,
            202,
            171,
            110,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            52,
            179,
            218,
            222,
            213,
            80,
            245,
            176,
            197,
            9,
            71,
            216,
            225,
            119,
            98,
            185,
            115,
            12,
            241,
            249,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            55,
            43,
            157,
            7,
            109,
            211,
            126,
            124,
            24,
            60,
            246,
            237,
            3,
            193,
            125,
            216,
            205,
            192,
            211,
            18,
          ],
          "type": "Buffer",
        },
        Object {
          "data": Array [
            110,
            53,
            133,
            177,
            23,
            81,
            66,
            58,
            114,
            235,
            91,
            73,
            245,
            121,
            64,
            190,
            80,
            144,
            71,
            87,
          ],
          "type": "Buffer",
        },
      ],
      "initRounds": 5,
    },
    "generatorPublicKey": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "height": 5,
    "id": Object {
      "data": Array [
        225,
        58,
        175,
        73,
        200,
        178,
        31,
        71,
        166,
        79,
        72,
        112,
        66,
        70,
        110,
        142,
        47,
        208,
        227,
        210,
        110,
        179,
        6,
        228,
        177,
        61,
        196,
        82,
        77,
        240,
        136,
        152,
      ],
      "type": "Buffer",
    },
    "previousBlockID": Object {
      "data": Array [
        69,
        70,
        144,
        161,
        195,
        120,
        56,
        50,
        96,
        7,
        81,
        154,
        124,
        225,
        200,
        166,
        164,
        149,
        223,
        80,
        137,
        143,
        30,
        189,
        105,
        210,
        47,
        188,
        237,
        249,
        104,
        154,
      ],
      "type": "Buffer",
    },
    "reward": 0n,
    "signature": Object {
      "data": Array [],
      "type": "Buffer",
    },
    "timestamp": 1591873718,
    "transactionRoot": Object {
      "data": Array [
        227,
        176,
        196,
        66,
        152,
        252,
        28,
        20,
        154,
        251,
        244,
        200,
        153,
        111,
        185,
        36,
        39,
        174,
        65,
        228,
        100,
        155,
        147,
        76,
        164,
        149,
        153,
        27,
        120,
        82,
        184,
        85,
      ],
      "type": "Buffer",
    },
    "version": 0,
  },
  "payload": Array [],
}
`;
